### PR TITLE
Piecewise form integrated with angular test

### PR DESCRIPTION
**Pull request by critzo**
**PR #100 continues the process of importing the Angular App with updated NDT test code, [addressing the second and third item in the planned development](https://github.com/opentechinstitute/piecewise/issues/92#issuecomment-267662413) for issue #92**  

**PR #100 should be reviewed after PR #99 has been reviewed and merged.**

**Summary of this PR**:
- Fully integrates Piecewise form fields to the Angular form, Piecewise form handling javascript functions to the Angular app, and connects the Angular form to Piecewise uwsgi collector
- Also beings to integrate the Piecewise map UI elements, but not fully.

**Testing**:
A successful test of this PR will be that after following the steps below, a user should be able to complete, the Piecewise web form elements, run an NDT test over HTTPS at ```https://<VM IP>/app/ ``` and form data should be submitted to the Piecewise database.

- [ ] Fork the repo, checkout this branch
- [ ] Add a hosts file to the repo with the contents: ```<VM IP> ansible_ssh_user=root```
- [ ] Set up a VM on Linode or elsewhere running Debian 8 (**you can use mine or I can set one up for you**)
- [ ] Log into the VM as root and run ```ssh-keygen```, then add your public SSH key to the root users' authorized_keys file in ```/root/.ssh/authorized_keys```
- [ ] Open the correct ports on your VM (the following were needed for a fresh Linode VM, ymmv):
```
root@debian:~# iptables -A INPUT -p tcp --dport 22 -j ACCEPT
root@debian:~# iptables -A INPUT -p tcp --dport 80 -j ACCEPT
root@debian:~# iptables -A INPUT -p tcp --dport 443 -j ACCEPT
```
- [ ] Return to your local repo and update the IP address for your VM in ```/conf/etc/nginx/sites-available/default```
- [ ] Update the git repo and version (branch) in ```/system_tasks.yml``` with the url for your fork
- [ ] Edit the file ```piecewise/piecewise/bigquery.py``` with a project ID that your google account has access to use. Your google account should also be subscribed to M-Lab Discuss
- [ ] Commit your changes to the branch on your fork, then run ```ansible-playbook -i hosts playbook.yml```
- [ ] Assuming the above completed with no errors, load the Angular app at ```https://<VM IP>/app```, and click past the security error for self-signed certs. You should see something like the attached pic: 
image
- [ ] SSH into your VM as root, change into the directory ```/opt/piecewise``` and run ```$ python -m piecewise.ingest``` -- you'll be prompted to authenticate the application with Google Oauth. **Complete the steps to authorize the app, but when the query begins, cancel using Ctrl-C** 
- [ ] Return to your root user home directory (```$ cd /root```) and download this [database dump](http://files.opentechinstitute.org/~critzo/pw.seattle.testing.db.tar.gz)
- [ ] Import the database dump into your Piecewise database using this command:
```$ gunzip -c pw.seattle.testing.db.tar.gz | pg_restore -U postgres -d piecewise -O --clean```
- [ ] Restart uwsgi: ```$ systemctl restart uwsgi ```
- [ ] Return to your browser, reload the page at ```https://<VM IP>/app``` and open your developer tools Console.
- [ ] Fill out the form and run the test. Once the test is completed, the Console should show a message "data submitted successfully". You can also confirm submission by logging into the VM, becoming the postgres user and querying the **extra_data** table directly.

---
This change is [Reviewable](https://reviewable.io/reviews/opentechinstitute/piecewise/100)

### CODE REVIEW

**Comment by nkinkade** [comment]
Hi Chris. I've added a few comments... largely around styling.

---

Reviewed 5 of 10 files at r1.
Review status: 5 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

*[reset_hosts.yml, line 4 at r1](https://reviewable.io:443/reviews/opentechinstitute/piecewise/100#-KZSy-LdSI5cQgqaWZDr:-KZSy-LdSI5cQgqaWZDs:b1d5xer) ([raw file](https://github.com/opentechinstitute/piecewise/blob/079e6ca523d91a92cd3eee6e454a9244de0bf508/reset_hosts.yml#L4)):*
> ```YAML
> - hosts: all
>   tasks:
>     - include: yml/reset_hosts.yml
> ```

What is the benefit of having an Ansible playbook in the root of the repo which includes more plays from a yml/ directory as opposed to just placing the contents of the included here, or just putting all playbooks in, for example, an ansible/ subdirectory or similar?

---

*[reset_hosts.yml, line 5 at r1](https://reviewable.io:443/reviews/opentechinstitute/piecewise/100#-KZSwt-eBckcUHND2WYU:-KZSwt-eBckcUHND2WYV:b6ptz4s) ([raw file](https://github.com/opentechinstitute/piecewise/blob/079e6ca523d91a92cd3eee6e454a9244de0bf508/reset_hosts.yml#L5)):*
> ```YAML
>   tasks:
>     - include: yml/reset_hosts.yml
>       become: True
> ```

I assume you've already tested this, and it may not matter, but the [documentation for `become`](http://docs.ansible.com/ansible/become.html) seems to asks for either `true` (lowercase T) or `yes`. Perhaps lowercasing the T here would be useful to distinguish this from some sort of constant?

---

*[system_tasks.yml, line 73 at r1](https://reviewable.io:443/reviews/opentechinstitute/piecewise/100#-KZT-zEbtjHz4Uj5rPfK:-KZT-zEbtjHz4Uj5rPfL:byycnu3) ([raw file](https://github.com/opentechinstitute/piecewise/blob/079e6ca523d91a92cd3eee6e454a9244de0bf508/system_tasks.yml#L73)):*
> ```YAML
>   command: openssl req -new -nodes -x509 -subj "/C=US/ST=District of Columbia/L=Washington/O=IT/CN=45.79.191.25" -days 3650 -out /etc/nginx/ssl/localcert.pem -keyout /etc/nginx/ssl/localcert.key
> - name: Enable nginx default site
>   file: src=/etc/nginx/sites-available/default dest=/etc/nginx/sites-enabled/default state=link
> ```

Where possible, in Ansible playbooks could we always use indention of action arguments for better readability?

```
file:
  src=/some/path
  dest=/another/path
  state=link
```

---

*[piecewise_web/app/index.html, line 14 at r1](https://reviewable.io:443/reviews/opentechinstitute/piecewise/100#-KZT0ohxzD3ubUOCn0Hm:-KZT0ohxzD3ubUOCn0Hn:blwwks1) ([raw file](https://github.com/opentechinstitute/piecewise/blob/079e6ca523d91a92cd3eee6e454a9244de0bf508/piecewise_web/app/index.html#L14)):*
> ```HTML
> 		<link rel="icon" href="/app/assets/images/mlab_logo_m.ico">
> 		<link href="https://api.tiles.mapbox.com/mapbox.js/v2.1.6/mapbox.css" rel="stylesheet" />
> 		<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/css/bootstrap.min.css">
> ```

For readability, maybe make sure the order of attributes is the same on all elements (e.g.,, href->rel)? 

---

*[piecewise_web/app/assets/css/main.css, line 1808 at r1](https://reviewable.io:443/reviews/opentechinstitute/piecewise/100#-KZT1REoN6gYkID8x7H_:-KZT1REoN6gYkID8x7Ha:b-8i92ek) ([raw file](https://github.com/opentechinstitute/piecewise/blob/079e6ca523d91a92cd3eee6e454a9244de0bf508/piecewise_web/app/assets/css/main.css#L1808)):*
> ```CSS
> 
> 	body {
> 		/* background: #312450; */
> ```

Throughout this file, where possible, I would recommend avoiding merely commenting things out, but instead just removed them fully. The change will be recorded in the history and it should make the file cleaner and less cluttered.

---

*[yml/reset_hosts.yml, line 6 at r1](https://reviewable.io:443/reviews/opentechinstitute/piecewise/100#-KZSyR_OJGBOgHaW3D6U:-KZSyR_OJGBOgHaW3D6V:b-uq217c) ([raw file](https://github.com/opentechinstitute/piecewise/blob/079e6ca523d91a92cd3eee6e454a9244de0bf508/yml/reset_hosts.yml#L6)):*
> ```YAML
>   with_items: 
>     - uwsgi
>     - nginx
> ```

nit: Add a blank line after this to visually separate this task from the next.

---

*[yml/reset_hosts.yml, line 11 at r1](https://reviewable.io:443/reviews/opentechinstitute/piecewise/100#-KZSyc5B6NRpoTuoKiss:-KZSyc5B6NRpoTuoKist:bebcxuc) ([raw file](https://github.com/opentechinstitute/piecewise/blob/079e6ca523d91a92cd3eee6e454a9244de0bf508/yml/reset_hosts.yml#L11)):*
> ```YAML
>     state: absent
> 
> - name: Purge package 
> ```

Could all of the "Purge package" tasks be grouped into a single task like the following?

```
- name: Purge packages
  apt:
    name: {{ item }}
    state: absent
  with_items:
    package0
    package1
    package2
    [...]
```

---

*[yml/reset_hosts.yml, line 19 at r1](https://reviewable.io:443/reviews/opentechinstitute/piecewise/100#-KZSzsK_ClGlMImTP_52:-KZSzsKavFX_Ib9SbUJ-:b-fs787h) ([raw file](https://github.com/opentechinstitute/piecewise/blob/079e6ca523d91a92cd3eee6e454a9244de0bf508/yml/reset_hosts.yml#L19)):*
> ```YAML
>   apt:  
>     name: postgis
>     name: postgresql
> ```

Is `state` missing here?

---

*[yml/reset_hosts.yml, line 66 at r1](https://reviewable.io:443/reviews/opentechinstitute/piecewise/100#-KZSzO3Nn6nqeOa8zg3Z:-KZSzO3Nn6nqeOa8zg3_:b-xgdxw5) ([raw file](https://github.com/opentechinstitute/piecewise/blob/079e6ca523d91a92cd3eee6e454a9244de0bf508/yml/reset_hosts.yml#L66)):*
> ```YAML
>     state: absent
> 
> - name: Remove Piecewise files
> ```

Could this be more cleanly rewritten like the following?

```
- name: Remove Piecewise files
  file:
    path: /opt/{{ item }}
    state: absent
  with_items:
    collector
    piecewise
    piecewise.git
    piecewise_web
    telescope
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opentechinstitute/piecewise/100#-:-KZT26vihVirAZ8_8CLT:b-mj4ugm)*

**Comment by critzo** [comment]
*[reset_hosts.yml, line 4 at r1](https://reviewable.io:443/reviews/opentechinstitute/piecewise/100#-KZSy-LdSI5cQgqaWZDr:-KZT3I5pXODTM3MTqoxL:bd6djd2) ([raw file](https://github.com/opentechinstitute/piecewise/blob/079e6ca523d91a92cd3eee6e454a9244de0bf508/reset_hosts.yml#L4)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

What is the benefit of having an Ansible playbook in the root of the repo which includes more plays from a yml/ directory as opposed to just placing the contents of the included here, or just putting all playbooks in, for example, an ansible/ subdirectory or similar?
</blockquote></details>

This was done for pure convenience for testing. Playbook.yml deploys to a host and this one tears it down. As for the current directory name, it's arbitrary and I plan to refactor everything Ansible later when approaching Issue #98 . If we want we can remove the yml sub-directory and associated files until that time to keep things cleaner.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opentechinstitute/piecewise/100)*

**Comment by critzo** [comment]
*[reset_hosts.yml, line 5 at r1](https://reviewable.io:443/reviews/opentechinstitute/piecewise/100#-KZSwt-eBckcUHND2WYU:-KZT45qYnQaRc70PSxXN:b-mhj1ke) ([raw file](https://github.com/opentechinstitute/piecewise/blob/079e6ca523d91a92cd3eee6e454a9244de0bf508/reset_hosts.yml#L5)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

I assume you've already tested this, and it may not matter, but the [documentation for `become`](http://docs.ansible.com/ansible/become.html) seems to asks for either `true` (lowercase T) or `yes`. Perhaps lowercasing the T here would be useful to distinguish this from some sort of constant?
</blockquote></details>

I have tested it and the case sensitivity doesn't matter to Ansible, but I can correct for internal consistency and distinguish it from constants like in JavaScript.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/opentechinstitute/piecewise/100)*

File: system_tasks.yml [hunk 1]
```diff
@@ -44,7 +44,7 @@
 - name: Fetch piecewise
   git: repo=https://github.com/opentechinstitute/piecewise.git 
        dest=/opt/piecewise.git/
-       version=import-angular-app
+       version=update-ndt-test
```
Comment (georgiamoon) [inline system_tasks.yml:47]: This should be master when merged right? the previous PR breaks the task because of the version issue.

**Comment by critzo** [comment]
I removed the in progress label and applied invalid, since this PR will no longer be necessary after the patch to NDTjs provided by @dwins , which will be applied in an upcoming PR.

**Pull request closed without merging**